File: 5.cpp
class Solution {
public:

    string longestPalindrome(string s) {
        pair<int,int> p = {0,0};
        vector<vector<int>> arr(s.length(), vector<int>(s.length(), 0));
        for(int i = 0; i < s.length(); i++){
            for(int j = 0; j <= i; j++){
                arr[i][j] = 1;
            }
        }
        for(int j = 1; j < s.length(); j++){
            int ii = 0;
            int jj = j;
            while(jj < s.length()){
                if(s[ii] == s[jj]){
                    arr[ii][jj] = arr[ii+1][jj-1];
                }
                if(arr[ii][jj] == 1 && abs(jj-ii) > abs(p.second - p.first)){
                    p.second = jj;
                    p.first = ii;
                }
                jj++;
                ii++;
            }
        }
        string sol = "";
        for(int i = p.first; i <= p.second; i++){
            sol = sol + s[i];
        }
        return sol;
    }
};

File: Python/5.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) == 1:
            return s
        dp = [[False] * len(s) for _ in range(len(s))]
        # print(dp)
        for i in range(len(s)):
            dp[i][i] = True
        # print(dp)
        for c in range(1, len(s)):
            i = 0
            j = c
            while i < len(s) and j < len(s):
                if s[i] == s[j]:
                    if j == i+1:
                        dp[i][j] = True
                    else:
                        dp[i][j] = dp[i+1][j-1]
                i += 1
                j += 1
        sol_len = 0
        sol_range = [-1,-1]
        # print(dp)
        for i in range(len(s)):
            for j in range(i, len(s)):
                if j-i+1 > sol_len and dp[i][j]:
                    sol_len = j-i+1
                    sol_range = [i,j]
        return s[sol_range[0]: sol_range[1]+1]
